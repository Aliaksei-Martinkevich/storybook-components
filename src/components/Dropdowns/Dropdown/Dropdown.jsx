import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import './Dropdown.scss';
import Header from './Header/Header.jsx';
import Content from './Content/Content.jsx';

class Dropdown extends Component {

  state = {
    isExpanded: false,
  };

  onHeaderClickHandler = () => {
    this.setState(prevState => ({ isExpanded: !prevState.isExpanded }));
  }

  onItemClickHandler = (itemId) => {
    this.props.onItemClick(itemId);
  }

  render() {
    return (
      <div className={classNames('dropdown', this.props.className, {
        dropdown_expanded: this.state.isExpanded,
        dropdown_collapsed: !this.state.isExpanded,
      },
      )}
      >
        <Header
          className="dropdown__header"
          onClick={this.onHeaderClickHandler}
          expanded={this.state.isExpanded}
        >
          {this.props.header}
        </Header>
        <Content
          expanded={this.state.isExpanded}
          className="dropdown__content"
          onItemClick={this.onItemClickHandler}
        >
          {this.props.children}
        </Content>
      </div>);
  }
}

Dropdown.defaultProps = {
  onItemClick: undefined,
  className: '',
};

Dropdown.propTypes = {
  header: PropTypes.element.isRequired,
  onItemClick: PropTypes.func,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]).isRequired,
  className: PropTypes.string,
};

export default Dropdown;
