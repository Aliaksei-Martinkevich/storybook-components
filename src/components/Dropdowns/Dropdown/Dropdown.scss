.dropdown {
    &__header {
        display: block;
    };
}