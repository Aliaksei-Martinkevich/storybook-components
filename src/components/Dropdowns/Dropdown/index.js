import React, { Component } from 'react';
import PropTypes from 'prop-types';
import './index.css';

import Headers from '../DropdownHeaders';
import Contents from '../DropdownContents';
const Header = Headers.Header;
const Content = Contents.Content;

class Dropdown extends Component { 

  constructor(props) {
    super(props);

    this.state = {
      isExpanded: false,
    }

    this.onClickHandler = this.onClickHandler.bind(this);
  }

  onClickHandler() {
    this.setState((prevState, props) => {
      return { isExpanded: !prevState.isExpanded };
    });
  }


  render() { 
    return (
      <div className={`dropdown ${this.state.isExpanded ? 'expanded' : 'collapsed'}`}>
        <Header onClick={ this.onClickHandler }>
          { this.props.header }
        </Header>
        <Content>
          { this.props.children }
        </Content>
      </div>)
  };
};

Dropdown.propTypes = {
  header: PropTypes.element,

}

export default Dropdown;
