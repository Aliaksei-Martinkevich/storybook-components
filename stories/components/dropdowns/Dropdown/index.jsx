import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';

import Dropdown from '../../../../src/components/Dropdowns/Dropdown/Dropdown.jsx';
import RegularItem from '../../../../src/components/Dropdowns/RegularItem/RegularItem.jsx';
import RegularHeader from '../../../../src/components/Dropdowns/RegularHeader/RegularHeader.jsx';
import ButtonHeader from '../../../../src/components/Dropdowns/ButtonHeader/ButtonHeader.jsx';
import InlineHeader from '../../../../src/components/Dropdowns/InlineHeader/InlineHeader.jsx';

export default storiesOf('Dropdown', module)
  .add('With regular items', () =>
    (<Dropdown onItemClick={action('Item clicked')} header="Plain text header">
      <RegularItem
        key="Item 1"
        text="Some text"
      />
      <RegularItem
        key="Item 2"
        text="Some long long long long long long long long long text"
      />
      <RegularItem
        key="Item 3"
        text="Some text with remark"
        remark="Some remark"
      />
      <RegularItem
        key="Item 4"
        text="Some text with remark"
        remark="Some long long long long long long long long long long long long two lines remark"
      />
      <RegularItem
        key="Item 5"
        text="👁 Some text with utf icon"
      />
      <RegularItem
        key="Item 6"
        text="Some text with icon"
        iconUri="https://material/io/guidelines/static/spec/images/callouts/default/svg"
      />
    </Dropdown>))
  .add('With regular header', () =>
    (<Dropdown
      onItemClick={action('Item clicked')}
      header={<RegularHeader text="Regular header" />}
    >
      <RegularItem
        text="Some text"
      />
      <RegularItem
        text="Some long long long long long long long long long text"
      />
      <RegularItem
        text="Some text with remark"
        remark="Some remark"
      />
      <RegularItem
        text="Some text with remark"
        remark="Some long long long long long long long long long long long long two lines remark"
      />
      <RegularItem
        text="👁 Some text with utf icon"
      />
      <RegularItem
        text="Some text with icon"
        iconUri="https://material/io/guidelines/static/spec/images/callouts/default/svg"
      />
    </Dropdown>))
  .add('With button header', () =>
    (<Dropdown
      onItemClick={action('Item clicked')}
      header={<ButtonHeader text="Button header" />}
    >
      <RegularItem
        text="Some text"
      />
      <RegularItem
        text="Some long long long long long long long long long text"
      />
      <RegularItem
        text="Some text with remark"
        remark="Some remark"
      />
      <RegularItem
        text="Some text with remark"
        remark="Some long long long long long long long long long long long long two lines remark"
      />
      <RegularItem
        text="👁 Some text with utf icon"
      />
      <RegularItem
        text="Some text with icon"
        iconUri="https://material/io/guidelines/static/spec/images/callouts/default/svg"
      />
    </Dropdown>))
  .add('With inline header', () =>
    (<Dropdown
      onItemClick={action('Item clicked')}
      header={<InlineHeader text="Inline header" />}
    >
      <RegularItem
        text="Some text"
      />
      <RegularItem
        text="Some long long long long long long long long long text"
      />
      <RegularItem
        text="Some text with remark"
        remark="Some remark"
      />
      <RegularItem
        text="Some text with remark"
        remark="Some long long long long long long long long long long long long two lines remark"
      />
      <RegularItem
        text="👁 Some text with utf icon"
      />
      <RegularItem
        text="Some text with icon"
        iconUri="https://material/io/guidelines/static/spec/images/callouts/default/svg"
      />
    </Dropdown>));
