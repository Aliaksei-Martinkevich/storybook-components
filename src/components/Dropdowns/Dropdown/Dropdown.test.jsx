import React from 'react';
import renderer from 'react-test-renderer';
import Dropdown from './Dropdown';

describe('Dropdown', () => {
  it('Should expands on click', () => {
    const customClassName = 'class-name-example';
    const Header = (<div>Header</div>);
    const onItemClick = jest.fn();
    const component = renderer.create(
      <Dropdown
        header={Header}
        className={customClassName}
        onItemClick={onItemClick}
      >
        <div>Item 1</div>
        <div>Item 2</div>
        <div>Item 3</div>
      </Dropdown>);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
    tree.children[0].props.onClick();
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();
    tree.children[1].children.forEach(elem => elem.props.onClick());
    expect(onItemClick.mock.calls).toEqual([['0'], ['1'], ['2']]);
  });
});
