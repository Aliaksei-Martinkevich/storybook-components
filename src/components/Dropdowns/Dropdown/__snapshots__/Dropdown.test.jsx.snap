// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown Should expands on click 1`] = `
<div
  className="dropdown class-name-example collapsed"
>
  <div
    className="dropdown-header dropdown__header"
    onClick={[Function]}
  >
    <div
      expanded={false}
    >
      Header
    </div>
  </div>
  <div
    className="dropdown-content dropdown__content collapsed"
  >
    <div
      className="dropdown-content__item"
      onClick={[Function]}
    >
      <div>
        Item 1
      </div>
    </div>
    <div
      className="dropdown-content__item"
      onClick={[Function]}
    >
      <div>
        Item 2
      </div>
    </div>
    <div
      className="dropdown-content__item"
      onClick={[Function]}
    >
      <div>
        Item 3
      </div>
    </div>
  </div>
</div>
`;

exports[`Dropdown Should expands on click 2`] = `
<div
  className="dropdown class-name-example expanded"
>
  <div
    className="dropdown-header dropdown__header"
    onClick={[Function]}
  >
    <div
      expanded={true}
    >
      Header
    </div>
  </div>
  <div
    className="dropdown-content dropdown__content expanded"
  >
    <div
      className="dropdown-content__item"
      onClick={[Function]}
    >
      <div>
        Item 1
      </div>
    </div>
    <div
      className="dropdown-content__item"
      onClick={[Function]}
    >
      <div>
        Item 2
      </div>
    </div>
    <div
      className="dropdown-content__item"
      onClick={[Function]}
    >
      <div>
        Item 3
      </div>
    </div>
  </div>
</div>
`;
